### PR TITLE
Return more information about named graphs

### LINKED ISSUE #644: LinkObj missing in Ontology Response
<http://localhost:3333/v2/ontologies/classes/http%3A%2F%2Fapi.knora.org%2Fontology%2Fknora-api%2Fv2%23LinkObj>

```
{"status":9,"error":"org.knora.webapi.NotFoundException: Some requested classes were not found: http://api.knora.org/ontology/knora-api/v2#LinkObj"}
```

### LINKED ISSUE #656: KnarQL query with incoming properties
Consider the following KnarQL query querying the book a page points to:

```
    PREFIX incunabula: <http://0.0.0.0:3333/ontology/incunabula/simple/v2#>
    PREFIX knora-api: <http://api.knora.org/ontology/knora-api/simple/v2#>
    
    CONSTRUCT {
        ?book knora-api:isMainResource true .
       
    } WHERE {
        ?book a knora-api:Resource .
        ?book a incunabula:book .
        
        <http://data.knora.org/8a9674d66e03> incunabula:partOf ?book .
        
        <http://data.knora.org/8a9674d66e03> a knora-api:Resource .
        
        incunabula:partOf knora-api:objectType knora-api:Resource .
    } 
```

The Knora API returns a `ForbiddenResource`.

Reason: the responder checks for the presence of the full graph pattern defined in the WHERE clause. But since book does not point to the page, the page cannot be contained in the book as a dependent resource (nested in one of its linking properties).

So there needs to be a way to represent "incoming" links as a virtual property, very much like a nested resource in a link value prop.

I suppose this has to happen in `ConstructResponseUtilV2`. Start with the main resource, but also check if there are incoming links (beware of circular references!).

### PR DESCRIPTION
**Pull request by tobiasschweizer**
- Return more information about named graphs, so the GUI can build an extended search form. Probably we should refactor `NamedGraphV1` to v2.
- add missing info to knora-api ontologies (simple and complex)
- Return incoming links using a virtual property: JSON-LD is a hierarchical structure, beginning with the main resource. Without a virtual incoming property, incoming resources cannot be contained in a nested structure:
     - find out about incoming resources in scala
     - nest incoming link values using a virtual property that does not actually exist in the triplestore
     - in case of an incoming link, swap take the `rdf:subject` as the source of the link (in other cases, the `rdf:object` is used to get the target)
     - adapt the way requested values are filtered out in the CONSTRUCT clause -> get variables representing resources (both for the main and dependent resources) and user-provided Iris representing resources

closes #644
closes #656

### CODE REVIEW

**Comment by tobiasschweizer** [comment]
@benjamingeer I added the Link Object to the Knora API ontologies in commit https://github.com/dhlab-basel/Knora/pull/651/commits/8e9bf3917713782ecccb22a25ae33bbd29e0bdea and fixed the tests.

Could you check if it is correct? 

Shouldn't link value props be mentioned in "hasOntologiesWithClasses"?

**Comment by benjamingeer** [comment]
Since `LinkObj` is a subclass of `Resource`, you need to manually include the cardinalities of `Resource` in the cardinalities of `LinkObj`. There is no cardinality inheritance for hard-coded classes.

> Shouldn't link value props be mentioned in "hasOntologiesWithClasses"?

Unless I've misunderstood, the answer is no, because properties are not classes...

**Comment by tobiasschweizer** [comment]
> Since LinkObj is a subclass of Resource, you need to manually include the cardinalities of Resource in the cardinalities of LinkObj. There is no cardinality inheritance for hard-coded classes.

Ok, I will do that!

> Shouldn't link value props be mentioned in "hasOntologiesWithClasses"?

Right! So it is fine.

**Comment by tobiasschweizer** [comment]
> Since LinkObj is a subclass of Resource, you need to manually include the cardinalities of Resource in the cardinalities of LinkObj. There is no cardinality inheritance for hard-coded classes.

done in https://github.com/dhlab-basel/Knora/pull/651/commits/49037c532b5f9f8885e0e79823273ce7d8aa30fe

**Comment by tobiasschweizer** [comment]
@benjamingeer I moved the JSOn test data files into a separate directory https://github.com/dhlab-basel/Knora/pull/651/commits/ed3cf640d6a68c1cf72beb3bdeb30ab958fcdb31

**Comment by tobiasschweizer** [comment]
@benjamingeer Dear Mr Geer: Could you have a look at https://github.com/dhlab-basel/Knora/pull/651/commits/c1aab0d258df19141192eeeefcdfd3b857e9a243 whenever you have a moment? Thanks!

**Comment by tobiasschweizer** [comment]
@SepidehAlassi https://github.com/dhlab-basel/Knora/pull/651/commits/9875b164981bcad24ba0569dfdb3e5bc20a40169 adds the standoff figure tag again which was deleted/overwritten by the last merge. But I still use it in the mapping for BEBB.

We should check which standoff tags are used in the mappings for Goldbach and BEBB and if they are used consistently. I will make an issue for that.

**Comment by tobiasschweizer** [comment]
@benjamingeer It would be great if you could resolve the merge conflicts with `develop` :-)

**Comment by benjamingeer** [comment]
There are definitely more merge conflicts on the way in #662. It might be less work to deal with them all at once. :)

**Comment by tobiasschweizer** [comment]
@benjamingeer Should we look at this PR together today? I would like to merge it asap.

**Comment by benjamingeer** [comment]
Yes, let’s do it this one and #662 this afternoon.

**Comment by tobiasschweizer** [comment]
@benjamingeer Ok!

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala [hunk 1]
```diff
@@ -383,14 +455,19 @@ object ConstructResponseUtilV2 {
                 IntervalValueContentV2(valueHasString = valueObjectValueHasString, valueHasIntervalStart = BigDecimal(valueObject.assertions(OntologyConstants.KnoraBase.ValueHasIntervalStart)), valueHasIntervalEnd = BigDecimal(valueObject.assertions(OntologyConstants.KnoraBase.ValueHasIntervalEnd)), comment = valueCommentOption)
 
             case OntologyConstants.KnoraBase.LinkValue =>
-                val referredResourceIri = valueObject.assertions(OntologyConstants.Rdf.Object)
+                val referredResourceIri = if (!valueObject.incomingLink) {
+                    valueObject.assertions(OntologyConstants.Rdf.Object)
+                } else {
+                    valueObject.assertions(OntologyConstants.Rdf.Subject)
+                }
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: Is this correct?

File: webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala [hunk 2]
```diff
@@ -455,29 +455,33 @@ object ConstructResponseUtilV2 {
...
+                        val nestedResourceIri = if (!valueObject.incomingLink) {
+                            targetResourceIri
+                        } else {
+                            sourceResourceIri
+                        }
 
                         // add information about the referred resource
                         linkValue.copy(
-                            referredResource = Some(constructReadResourceV2(referredResourceIri, referredResourceAssertions, mappings)) // construct a `ReadResourceV2`
+                            referredResource = Some(constructReadResourceV2(nestedResourceIri, nestedResourceAssertions, mappings)) // construct a `ReadResourceV2`
```
Comment (benjamingeer) [inline webapi/src/main/scala/org/knora/webapi/util/ConstructResponseUtilV2.scala]: Should this still be called `referredResource`?

**Pull request merged**